seq2simulate/run_make_proviral_data.py

```diff
@@ -0,0 +1,483 @@
+    return fastq_files, sam_files
+
+
+def _open_all_data_files(sequences, fastq_files, sam_files, diff_files):
+    open_fq_files = {}
+    for mutation_type, fq in fastq_files.iteritems():
+        fastq_siblings = []
+        for _fq in fq:
+            with open(_fq, 'r') as f:
+                fastq_siblings.append([i for i in parse_fastq(f)])
+        open_fq_files[mutation_type] = fastq_siblings
+    
+    open_sam_files = {}
+    for mutation_type, fs in sam_files.iteritems():
+        with open(fs, 'r') as f:
+            sam_reads = {}
+            for sam_read in parse_sam(f):
+                read_id = sam_read[0].strip('\n') 
+                if read_id in sam_reads:
+                    sam_reads[read_id].append(sam_read[1:])
+                else:
+                    sam_reads[read_id] = [sam_read[1:]]
+            open_sam_files[mutation_type] = sam_reads
+
+    open_diff_files = {'null': {i.id: [] for i in sequences}}
+    for mutation_type, fd in diff_files.iteritems():
+        with open(fd, 'r') as f:
+            open_diff_files[mutation_type] = pickle.load(f)
+    return open_fq_files, open_sam_files, open_diff_files
+
+
+def _decorate_fastq_headers(fastq_sample, mutation_type, sam_file=None, diff_file=None, paired_end=False):
+
+    """
+    Decorate FASTQ headers with a suffix code indicating whether or not this
+    read covers a known mutation (see MUTATIONS for mutation codes).
+    """
+    for i_read, read in enumerate(fastq_sample):
+        id_suffix = None
+        if mutation_type != 'null':
+            if sam_file is not None and diff_file is not None:
+                read_id = read[0][1:].strip('\n')
+                if paired_end:
+                    read_id, read_direction = read_id[:-2], read_id[-1]
+                    sam_line_f, sam_line_r = _parse_sam_line(read_id, sam_file, paired_end=paired_end)
+                    if read_direction == '1':
+                        sam_line = sam_line_f
+                    elif read_direction == '2':
+                        sam_line = sam_line_r
+                    else:
+                        raise ValueError('Paired-end FASTQ read headers must have directionality expressed as .../1 or .../2')
+                else:
+                    sam_line = _parse_sam_line(read_id, sam_file, paired_end=paired_end)
+                seq_diffs = diff_file[sam_line['seq_id']]
+                id_suffix = _get_id_suffix(mutation_type, sam_line, seq_diffs)
+        if id_suffix is None:
+            id_suffix = '_{}\n'.format(MUTATIONS['null'])
+            
+        read[0] = read[0][:-1] + id_suffix
+
+        fastq_sample[i_read] = ''.join(read)
+
+def _get_id_suffix(mutation_type, sam_line, seq_diffs):
+
+    critical_num_hypermutations = 6
+
+    id_suffix = None
+    if mutation_type == 'hypermutation' and _is_hypermutated(sam_line, seq_diffs, critical_num_hypermutations):
```

Comment:
again, lots of cut-and-paste, needs to be compressed into a different structure.